test/foundry/src/concrete/vault/OffchainAssetReceiptVault.handler.t.sol

```diff
@@ -0,0 +1,216 @@
+// SPDX-License-Identifier: CAL
+pragma solidity =0.8.25;
+
+import {
+    OffchainAssetReceiptVault,
+    OffchainAssetReceiptVaultConfig
+} from "../../../../../contracts/concrete/vault/OffchainAssetReceiptVault.sol";
+import {OffchainAssetReceiptVaultTest, Vm} from "test/foundry/abstract/OffchainAssetReceiptVaultTest.sol";
+import {LibOffchainAssetVaultCreator} from "test/foundry/lib/LibOffchainAssetVaultCreator.sol";
+import {Receipt as ReceiptContract} from "../../../../../contracts/concrete/receipt/Receipt.sol";
+
+contract OffchainAssetReceiptVaultHandlerTest is OffchainAssetReceiptVaultTest {
+    event SetERC20Tier(address sender, address tier, uint256 minimumTier, uint256[] context, bytes data);
+    event SetERC1155Tier(address sender, address tier, uint256 minimumTier, uint256[] context, bytes data);
+    event OffchainAssetReceiptVaultInitialized(address sender, OffchainAssetReceiptVaultConfig config);
+    event DepositWithReceipt(
+        address sender, address owner, uint256 assets, uint256 shares, uint256 id, bytes receiptInformation
+    );
+
+    /// Get Receipt from event
+    function getReceipt() internal returns (ReceiptContract) {
+        Vm.Log[] memory logs = vm.getRecordedLogs();
+
+        // Find the OffchainAssetReceiptVaultInitialized event log
+        address receiptAddress = address(0);
+        bool eventFound = false; // Flag to indicate whether the event log was found
+        for (uint256 i = 0; i < logs.length; i++) {
+            if (logs[i].topics[0] == OffchainAssetReceiptVaultInitialized.selector) {
+                // Decode the event data
+                (, OffchainAssetReceiptVaultConfig memory config) =
+                    abi.decode(logs[i].data, (address, OffchainAssetReceiptVaultConfig));
+                receiptAddress = config.receiptVaultConfig.receipt;
+                eventFound = true; // Set the flag to true since event log was found
+                break;
+            }
+        }
+
+        // Assert that the event log was found
+        assertTrue(eventFound, "OffchainAssetReceiptVaultInitialized event log not found");
+        // Return an receipt contract
+        return ReceiptContract(receiptAddress);
+    }
+
+    /// Test testReceiptTransfer to self with handler role
+    function testReceiptTransferHandler(
+        uint256 fuzzedKeyAlice,
+        uint256 fuzzedKeyBob,
+        string memory assetName,
+        uint256 referenceBlockNumber,
+        uint256 certifyUntil,
+        uint256 futureTimeStamp,
+        bool forceUntil,
+        uint256 balance
+    ) external {
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        address alice = vm.addr((fuzzedKeyAlice % (SECP256K1_ORDER - 1)) + 1);
+        address bob = vm.addr((fuzzedKeyBob % (SECP256K1_ORDER - 1)) + 1);
+
+        balance = bound(balance, 1, type(uint256).max); // Bound from one to avoid ZeroAssets
```

Comment:
this feels like a lot of boilerplate used across all these tests

the tests are largely the same, but there's a lot of repeated code, it would be better to make an abstraction so that only the key differences remain in each test